### PR TITLE
Reduce lock contention for metrics hashName hasher

### PR DESCRIPTION
**Pull request by shanemhansen**
The singleton hasher used by hashName can be highly contented and reduce overall multi-core throughput. This has been observed in applications that dynamically create lots of counters. This change adds a parallel benchmark to measure multi-core performance for applications. It also changes the singleton hasher to use sync.Pool and thus benefit from it's per-P cache:

- https://www.sobyte.net/post/2022-03/think-in-sync-pool/

In benchmarks this change can provide ~4x performance improvements in multicore situations.

### CODE REVIEW

File: sdks/go/pkg/beam/core/metrics/metrics_test.go [hunk 1]
```diff
@@ -617,21 +617,23 @@ func TestPcolQueryResult(t *testing.T) {
 	}
 }
 
-// Run on @lostluck's desktop (2020/01/21) go1.13.4
+// Run on @shanemhansen's desktop (2022/01/03) go1.20 RC1 after changing hashName to use a pool of hashers
+// sync.Pool can return thread-local results eliminating the need for a lock and increasing throughput.
+// There are users in the wild who create an excessive number of Counters so a 4x improvement in throughput at the expense of
+// creating GOMAXPROCS hasher values seems reasonable.
```
Comment (damccorm) [inline sdks/go/pkg/beam/core/metrics/metrics_test.go:623]: We should probably cut some of this before merging.

File: sdks/go/pkg/beam/core/metrics/metrics_test.go [hunk 2]
```diff
@@ -617,21 +617,23 @@ func TestPcolQueryResult(t *testing.T) {
...
-// BenchmarkMetrics/gauge_predeclared-12                   18535839                60.5 ns/op             0 B/op          0 allocs/op
-// BenchmarkMetrics/counter_raw-12                         159581343                7.18 ns/op            0 B/op          0 allocs/op
-// BenchmarkMetrics/distribution_raw-12                    82724314                14.7 ns/op             0 B/op          0 allocs/op
-// BenchmarkMetrics/gauge_raw-12                           23292386                55.2 ns/op             0 B/op          0 allocs/op
-// BenchmarkMetrics/getStore-12                            309361303                3.78 ns/op            0 B/op          0 allocs/op
-// BenchmarkMetrics/getCounterSet-12                       287720998                3.98 ns/op            0 B/op          0 allocs/op
+// name                                 old time/op  new time/op  delta
+// Metrics/counter_inplace-12            376ns ±17%    88ns ± 7%  -76.66%  (p=0.008 n=5+5)
+// Metrics/distribution_inplace-12       394ns ± 3%   153ns ± 8%  -61.17%  (p=0.008 n=5+5)
+// Metrics/gauge_inplace-12              371ns ± 4%   258ns ± 1%  -30.37%  (p=0.008 n=5+5)
```
Comment (damccorm) [inline sdks/go/pkg/beam/core/metrics/metrics_test.go:628]: Nice

**Review by damccorm — APPROVED** [review]
This looks good to me, but I'd like to get a second set of eyes with deeper Go performance expertise - @lostluck could you take a look?

File: sdks/go/pkg/beam/core/metrics/metrics_test.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/core/metrics/metrics_test.go:623]: I'm OK with keeping it. It's pedagogical.

**Comment by lostluck** [comment]
It's been a while since I looked at sync.Pool, so there are probably other similar opportunities available (WRT general object re-use) elsewhere in the SDK. More often than not, each bundle gets its own instance of everything it needs, but there are one or two common resource locks we might be able to do better with.

**Pull request merged**